library/core/src/ptr/mut_ptr.rs

```diff
@@ -2196,4 +2204,67 @@ impl<T: ?Sized> PartialOrd for *mut T {
     fn ge(&self, other: &*mut T) -> bool {
         *self >= *other
     }
+
+}
+
+#[cfg(kani)]
+#[unstable(feature = "kani", issue = "none")]
+mod verify {
+    use crate::kani;
+    use kani::{PointerGenerator};
+
+    #[kani::proof_for_contract(<*mut ()>::offset_from)]
+    pub fn check_mut_offset_from_unit() {
+        let mut val: () = ();
+        let src_ptr: *mut () = &mut val;
+        let dest_ptr: *mut () = &mut val;
+        unsafe {
+            dest_ptr.offset_from(src_ptr);
+        }
+    }
+
+    macro_rules! generate_offset_from_harness {
+        ($type: ty, $proof_name: ident) => {
+            #[kani::proof_for_contract(<*mut $type>::offset_from)]
+            pub fn $proof_name() {
+                let mut val: $type = kani::any::<$type>();
+                let ptr: *mut $type = &mut val;
+
+                let offset: usize = kani::any_where(|x| *x <= 1);
+                let src_ptr: *mut $type = unsafe { ptr.add(offset) };
+
+                let offset: usize = kani::any_where(|x| *x <= 1);
+                let dest_ptr: *mut $type = unsafe { ptr.add(offset) };
```

Comment:
Can we non-deterministically decide the provenance of the dest_ptr and generate it accordingly?